torch/distributed/rpc/api.py

```diff
@@ -235,6 +236,7 @@ def shutdown(graceful=True):
     """
     if graceful:
         _wait_all_workers()
+        _delete_all_user_rrefs()
         _get_current_rpc_agent().join()
     try:
         # This raises a `TORCH_CHECK()` exception on RRef leak detected.
```

Comment:
to be more precise, we need to move the line below (`_destroy_rref_context `) to the front of `_get_current_rpc_agent().join()`. And we probably can also consolidate `_destroy_rref_context` and `_delete_all_user_rrefs` to one function, e.g., `wait_and_clear_rref_context`.

Reply:
The problem is that I didn't wait for response of delete user rref.
So user could run to `_agent.join()` right after sending delete user rref request, while owner is also waiting on `_agent.join()`. At this time, owner can destroy rref context and report owner rref leaks. After this, the server thread picks up the request to delete user rref and realized no such rref id exists and the owner node crashes. Then, User node detects GLOO connection cloed by peer.

Reply:
Keeping it as is, cause we don't want to mix graceful and non-graceful together. Non-graceful should not wait on anything.